DQA-8115: Component check - Forbid deprecated configuration.

mike-michal opened the pull request:
Make sure that the forbidden/obsolete file is not present in the project.

mike-michal committed:
DQA-8115: (and DQA-8116) Make sure that the forbidden/obsolete file is not present in the project.

mike-michal committed:
DQA-8115: (and DQA-8116) PhpCS

mike-michal committed:
DQA-8115: (and DQA-8116) Suppress ExcessiveMethodLength

mike-michal committed:
DQA-8115: (and DQA-8116) Check if the file must be forbidden under the condition.

mike-michal committed:
DQA-8115: (and DQA-8116) PhpCS

mike-michal committed:
DQA-8115: (and DQA-8116) PhpUnit Test fix.

mike-michal committed:
DQA-8118: Detect deprecated env variables (like: ASDA_USER, ASDA_PASSWORD, QA_API_BASIC_AUTH).

mike-michal committed:
DQA-8118: Removed deprecated env variables.

mike-michal committed:
DQA-8118: PhpUnit test added.

joaocsilva commented on src/Helpers/ProjectInfo.php:
```diff
@@ -0,0 +1,30 @@
+     */
+    public function getEnabledDrupalModules() {
+        $enabledModules = [];
+        $result = $this->taskExec($this->getBin('drush') . ' pm-list --fields=status --format=json')
```
What do you think of using the `pm:list` command option `status`, the command could be:
```
drush pm-list --status=enabled --format=json
```

https://www.drush.org/12.x/commands/pm_list/

mike-michal replied:
Hi @joaocsilva this has been fixed, please proceed with the review.

joaocsilva commented on src/Callbacks/ForbiddenFilesConditions.php:
```diff
@@ -0,0 +1,21 @@
+{
+
+    public static function grumpPhpisNotInstalled() {
+        // If grumphp package is not present in a project, then grumphp config file must not be present.
```
This comment can be moved into the method Docblock

mike-michal replied:
Hi @joaocsilva this has been fixed.

joaocsilva commented on src/TaskRunner/Commands/ComponentCheckCommands.php:
```diff
@@ -172,6 +175,39 @@ protected function prepareSkips(): void
+            }
+        }
+
+        if (!$this->composerFailed) {
```
If this is a new section in the component-check, I will suggest to have its own flag like `$this->configurationFailed` to avoid misunderstanding.

mike-michal replied:
Hi @joaocsilva this has been fixed.

joaocsilva reviewed (commented):
Hi @ozyrys ,
Please have a look into my comments, the overall looks good but I have some suggestions.
Best regards,
João Silva

mike-michal committed:
DQA-8118: Fix.

mike-michal committed:
DQA-8118: Fix.

mike-michal committed:
DQA-8118: Removed outdated environment variable.

mike-michal committed:
DQA-8118: Removed outdated environment variable.

mike-michal committed:
DQA-8118: TODO: Move it to AbstractCommands.php::getDockerComposeYml.

mike-michal committed:
DQA-8115: (and DQA-8116) Fix.

mike-michal committed:
DQA-8115: (and DQA-8116) Forgotten file for the previous commit.

mike-michal committed:
DQA-8115: (and DQA-8116) Get enabled modules - back to previous version + drush command performance improvement.

mike-michal committed:
DQA-8115: (and DQA-8116) Comment fix.

mike-michal committed:
DQA-8115: (and DQA-8116) PhpCS.

mike-michal committed:
DQA-8118: AbstractCommands improvement - encapsulated getYml function + redundancy fixed - getComposerJson and getComposerLock functions unified into getJson.

mike-michal committed:
DQA-8118: Moved const DC_YML_FILE to be available in ComponentCheckCommands.

mike-michal committed:
DQA-8118: Reverted "encapsulated getYml function" - it's not worth it.

mike-michal committed:
DQA-8118: PhpCS.

mike-michal committed:
DQA-8118: Detect deprecated env variables in .env and .env.dist files as well.

mike-michal committed:
DQA-8118: Detect if deprecated env variables are not already present in the container.

mike-michal committed:
DQA-8118: Performance improvement on deprecated env variables detection.

mike-michal committed:
DQA-8115: (and DQA-8116) Fix for the Condition Callbacks, using call_user_func_array function for static methods instead of creating an instance of a class to check if method exists.

mike-michal committed:
DQA-8118: Removed deprecated env variables from Drone pipeline.

mike-michal committed:
DQA-8118: Deprecated ASDA test removed, please adapt the Toolkit code to deprecated and removed ASDA type -> DQA-8510.

mike-michal committed:
Merge branch 'feature/DQA-8118' into feature/DQA-8115

* feature/DQA-8118:
  DQA-8118: Deprecated ASDA test removed, please adapt the Toolkit code to deprecated and removed ASDA type -> DQA-8510.
  DQA-8118: Removed deprecated env variables from Drone pipeline.
  DQA-8118: Performance improvement on deprecated env variables detection.
  DQA-8118: Detect if deprecated env variables are not already present in the container.
  DQA-8118: Detect deprecated env variables in .env and .env.dist files as well.
  DQA-8118: PhpCS.
  DQA-8118: Reverted "encapsulated getYml function" - it's not worth it.
  DQA-8118: Moved const DC_YML_FILE to be available in ComponentCheckCommands.
  DQA-8118: AbstractCommands improvement - encapsulated getYml function + redundancy fixed - getComposerJson and getComposerLock functions unified into getJson.
  DQA-8118: TODO: Move it to AbstractCommands.php::getDockerComposeYml.
  DQA-8118: Removed outdated environment variable.
  DQA-8118: Removed outdated environment variable.
  DQA-8118: Fix.
  DQA-8118: Fix.
  DQA-8118: PhpUnit test added.
  DQA-8118: Removed deprecated env variables.
  DQA-8118: Detect deprecated env variables (like: ASDA_USER, ASDA_PASSWORD, QA_API_BASIC_AUTH).

# Conflicts:
#	src/TaskRunner/Commands/ComponentCheckCommands.php
#	tests/fixtures/commands/component-check.yml

joaocsilva commented on config/runner/toolkit.yml:
```diff
@@ -15,6 +15,12 @@ toolkit:
     mydumper: '/scripts/backupdb.sh'
+  forbidden_files:
+    - name: ".env"
+      error: ".env file detected!\n In order to avoid loading .env file in ACC or PROD environments, please rename it to something like .env.dist\n The file belongs to your application. You should guarantee that is not being called into any point of your Drupal site. If you are only calling it in docker-compose.yml, it's recommended to drop entirely the file and declare the environment variables  ...
```
Hi @ozyrys , the correct filename is `docker-compose.yml`

mike-michal replied:
Hi @joaocsilva it's fixed.

joaocsilva commented on config/runner/toolkit.yml:
```diff
@@ -15,6 +15,12 @@ toolkit:
+    - name: ".env"
+      error: ".env file detected!\n In order to avoid loading .env file in ACC or PROD environments, please rename it to something like .env.dist\n The file belongs to your application. You should guarantee that is not being called into any point of your Drupal site. If you are only calling it in docker-compose.yml, it's recommended to drop entirely the file and declare the environment variables  ...
+    - name: "grumphp.yml.dist"
+      error: "grumphp.yml.dist file detected!\n The usage of GrumPHP was removed in the upgrade for ec-europa/toolkit:^9, please remove grumphp.yml.dist file."
```
I would say this in a different way.
Something like: 

The grumphp.yml.dist file was detected but the package is not being required by your project, please remove the file.

I think is not important to know when it was removed, but instead, would be good to tell them that they have configuration file that is not being used.

mike-michal replied:
Hi @joaocsilva it's fixed.

joaocsilva commented on src/TaskRunner/Commands/ComponentCheckCommands.php:
```diff
@@ -172,6 +176,39 @@ protected function prepareSkips(): void
+            // Check if the file must be forbidden under the condition.
+            if (!empty($file['condition_callback']) && is_callable($file['condition_callback'])) {
+                // Invoke the callback method and if the condition is not met then don't forbid the file.
+                if (call_user_func_array($file['condition_callback'], []) === false) {
```
If the parameters are not used, could this function be used instead ?
https://www.php.net/manual/en/function.call-user-func.php

joaocsilva commented on tests/fixtures/commands/component-check.yml:
```diff
@@ -307,6 +336,8 @@
           drupal_patches: false
   tokens: ''
   resources:
+    - from: 'sample-docker-default.yml'
```
Do we need to always provide this file ?

mike-michal replied:
Hi @joaocsilva it's fixed - docker-compose.yml has been removed from all the tests. 
  Apparently, we don't need it. It must have been added when functionality was not fully ready. Good catch, thank you.

joaocsilva commented on src/TaskRunner/Commands/ComponentCheckCommands.php:
```diff
@@ -227,6 +264,54 @@ protected function componentComposer()
+                $ext = pathinfo($filename, PATHINFO_EXTENSION);
+                if ($ext && $ext == 'yml') {
+                    $setEnvVars[$filename] = Yaml::parseFile($filename);
+                    // TODO: If service "web" has different name it will not work. Think of something dynamic.
```
I do agree, lets do a loop and check if `environment` key exists.

joaocsilva reviewed (changes requested):
Hi @ozyrys 
Please have a look to my comments

joaocsilva committed:
Merge branch 'release/10.x' into feature/DQA-8115

mike-michal committed:
DQA-8115: (and DQA-8116) Fixed typo and description.

mike-michal committed:
DQA-8115: (and DQA-8116) Substituting with a more lightweight function.

mike-michal committed:
DQA-8115: (and DQA-8116) Loop through all the Docker services looking for forbidden environment variables.

mike-michal committed:
DQA-8115: (and DQA-8116) Removed useless file from PhpUnit tests.

mike-michal committed:
DQA-8115: (and DQA-8116) PhpCS

mike-michal committed:
DQA-8115: Merge branch 'release/10.x' into feature/DQA-8115

# Conflicts:
#	config/runner/toolkit.yml
#	tests/fixtures/commands/component-check.yml

joaocsilva commented on src/TaskRunner/Commands/ComponentCheckCommands.php:
```diff
@@ -248,6 +285,62 @@ protected function componentComposer()
+            // Check if forbidden env variables are not already here.
+            if (getenv($varName) !== false) {
+                $this->configurationFailed = true;
+                $this->io->error('Fiorbidden environment variable "' . $varName . '" detected in the container. Please locate the source of that variable and remove it.');
```
Hi @ozyrys ,
I just noticed this typo here `Fiorbidden`.

joaocsilva commented on src/TaskRunner/Commands/ComponentCheckCommands.php:
```diff
@@ -248,6 +285,62 @@ protected function componentComposer()
+                foreach ($envVarsSet as $filename => $envVars) {
+                    if (array_key_exists($varName, $envVars)) {
+                        $this->configurationFailed = true;
+                        $this->io->error('Fiorbidden environment variable detected in ' . $filename . ' file: ' . $varName . '. Please remove it.');
```
Also here

joaocsilva commented on tests/fixtures/commands/component-check.yml:
```diff
@@ -69,20 +69,27 @@
+         Evaluation module check                  failed
+         Development module check                 passed
+         Composer validation check                passed
+         Project configuration validation check   passed
```
I am wondering, can we short this label ?
Something like `Project configuration check`

jonhy81 commented on src/TaskRunner/Commands/DockerCommands.php:
```diff
@@ -54,8 +53,8 @@ public function dockerRefreshConfiguration(): int
             $this->say("The file $dockerCompose was not found, creating it.");
```
Change to use of constant

mike-michal committed:
DQA-8115: (and DQA-8116) Moved docker compose file name constant back into its commands class + Fixed typos and being more concise with the labels.

mike-michal committed:
DQA-8115: (and DQA-8116) Fix for PhpUnit test.

joaocsilva committed:
Merge branch 'release/10.x' into feature/DQA-8115

joaocsilva reviewed (approved):
Hi @ozyrys ,
Looks good
